File: Algorithms/LinkedList/doublyList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.sz = 0

    def size(self):
        # Returns the sz of the list
        return self.sz

    def is_empty(self):
        # Returns True if the list is empty, False otherwise
        return self.head is None

    def append(self, data):
        # Adds a node with the given data to the end of the list
        new_node = Node(data)
        if self.is_empty():
            self.head = self.tail = new_node
        else:
            self.tail.next = new_node
            new_node.prev = self.tail
            self.tail = new_node
        self.sz += 1

    def prepend(self, data):
        # Adds a node with the given data to the start of the list
        new_node = Node(data)
        if self.is_empty():
            self.head = self.tail = new_node
        else:
            self.head.prev = new_node
            new_node.next = self.head
            self.head = new_node
        self.sz += 1

    def delete(self, data):
        # Deletes the first occurrence of a node with the given data
        cur = self.head
        while cur:
            if cur.data == data:
                if cur.prev:
                    cur.prev.next = cur.next
                else:
                    self.head = cur.next

                if cur.next:
                    cur.next.prev = cur.prev
                else:
                    self.tail = cur.prev

                return
            cur = cur.next

    def find(self, data):
        # Returns True if data is in the list, False otherwise
        cur = self.head
        while cur:
            if cur.data == data:
                return True
            cur = cur.next
        return False

    def print_list(self):
        # Prints the list in forward direction
        cur = self.head
        while cur:
            print(cur.data, end=' ')
            cur = cur.next
        print()


# Example Usage
dll = DoublyLinkedList()
dll.append(1)
dll.append(2)
dll.append(3)
dll.prepend(0)
print("List:", end=' ')
dll.print_list()

print("Length:", dll.size())
print("Is 2 in the list? : ", dll.find(2))

dll.delete(2)
print("List after deleting 2: ", end=' ')
dll.print_list()

File: Algorithms/LinkedList/doublyList.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long

class MyList
{
private:
    struct node
    {
        int val;
        node *next;
        node *prev;
        node(int val)
        {
            this->val = val;
            this->next = nullptr;
            this->prev = nullptr;
        }
    };
    int sz = 0;
    node *head;
    node *tail;

public:
    MyList()
    {
        sz = 0;
        head = nullptr;
        tail = nullptr;
    }

    void push_front(int val)
    {
        node *newNode = new node(val);
        newNode->next = head;
        if (sz != 0)
            head->prev = newNode;

        head = newNode;

        if (sz == 0)
            tail = head;
        sz++;
    }

    void push_back(int val)
    {
        if (sz == 0)
        {
            push_front(val);
            return;
        }
        node *newNode = new node(val);
        tail->next = newNode;
        newNode->prev = tail;
        tail = newNode;
        sz++;
    }

    void insert_front(int posi, int val)
    {
        if (posi < 0 || posi >= sz)
            return;
        if (posi == 0)
        {
            push_front(val);
            return;
        }

        node *cur = head;
        for (int i = 0; i < (posi - 1); i++)
        {
            cur = cur->next;
        }

        node *newNode = new node(val);
        newNode->next = cur->next;
        newNode->prev = cur;
        cur->next = newNode;
        newNode->next->prev = newNode;
        sz++;
    }

    void insert_back(int posi, int val)
    {
        if (posi < 0 || posi >= sz)
            return;
        if (posi == 0)
        {
            push_back(val);
            return;
        }
        node *cur = tail;
        for (int i = 0; i < (posi - 1); i++)
        {
            cur = cur->prev;
        }
        node *newNode = new node(val);
        newNode->prev = cur->prev;
        newNode->next = cur;
        newNode->prev->next = newNode;
        sz++;
    }

    void erase_front(int posi)
    {
        if (posi < 0 || posi >= sz)
            return;

        node *delNode;
        if (posi == 0)
        {
            delNode = head;
            head = head->next;
            if (head)
                head->prev = nullptr;
        }
        else
        {
            node *cur = head;
            for (int i = 0; i < (posi - 1); i++)
            {
                cur = cur->next;
            }
            delNode = cur->next;
            cur->next = cur->next->next;
            if (posi != (sz - 1))
                cur->next->prev = cur;
        }
        delete delNode;
        sz--;
    }
    void erase_back(int posi)
    {
        if (posi < 0 || posi >= sz)
            return;

        node *delNode;
        if (posi == 0)
        {
            delNode = tail;
            tail = tail->prev;
            if (tail)
                tail->next = nullptr;
        }
        else
        {
            node *cur = tail;
            for (int i = 0; i < (posi - 1); i++)
            {
                cur = cur->prev;
            }
            delNode = cur->prev;
            cur->prev = cur->prev->prev;
            if (posi != (sz - 1))
                cur->prev->next = cur;
        }
        delete delNode;
        sz--;
    }

    void print()
    {
        node *cur = head;
        while (cur != nullptr)
        {
            cout << cur->val << " ";
            cur = cur->next;
        }
        cout << endl;
    }

    ll size(){
        return sz;
    }
};

signed main()
{
    MyList A;
    A.push_back(0);
    A.push_back(1);
    A.push_back(2);
    A.push_back(3);
    A.print();
    A.erase_back(0);
    A.print();
    A.erase_front(1);
    A.print();

    return 0;
}